### PR TITLE
Add google_fonts package to experimental

### PR DESCRIPTION
**Pull request by clocksmith**
Files needing review are inside google_fonts/lib (google_fonts.dart is generated)

### CODE REVIEW

File: google_fonts/README.md [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+![](main.gif)
+
+With the GoogleFonts package, ttf files do not need to be stored in your assets folder and mapped in
+the pubspec. Instead, they are fetched once via http at runtime, and stored in the apps local disk
+space. This is ideal for development, and can be the preferred behavior for production apps that
+are looking to save space in the initial app bundle size.
+
+To use google fonts with the default TextStyle:
+
+```
```
Comment (jacob314) [inline google_fonts/README.md]: should this be
"```dart"

?

**Comment by clocksmith** [comment]
Done, I left fonts_data.json (so you can see the structure that the
generator parses from). It is just a downloaded file from fonts.google.com

File: google_fonts/lib/src/README.md [hunk 2]
```diff
@@ -0,0 +1,4 @@
+To generate the GoogleFonts class:
+
+1. Navigate to the root directory of this project
```
Comment (jacob314) [inline google_fonts/lib/src/README.md]: Are there extra steps?

File: google_fonts/lib/src/generator.dart [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'dart:convert';
+import 'dart:io';
+
+import 'package:mustache/mustache.dart';
+
+void main() {
+  final fontsJsonData = readFontsJsonData();
```
Comment (jacob314) [inline google_fonts/lib/src/generator.dart]: Add a TODO to integrate with package:build
That would at some point let you avoid checking in the generated files.

File: google_fonts/lib/src/google_fonts_base.dart [hunk 4]
```diff
@@ -0,0 +1,190 @@
...
+
+import 'dart:io';
+import 'dart:typed_data';
+import 'dart:ui';
+import 'package:flutter/material.dart';
+import 'package:http/http.dart' as http;
+import 'package:flutter/services.dart';
+import 'package:path_provider/path_provider.dart';
+
+Set<String> _loadedFonts = {};
```
Comment (jacob314) [inline google_fonts/lib/src/google_fonts_base.dart]: nit: make this final.

File: google_fonts/lib/src/google_fonts_base.dart [hunk 5]
```diff
@@ -0,0 +1,190 @@
...
+
+  _loadedFonts.add(fontName);
+  final fontLoader = FontLoader(fontName);
+  var byteData = _readLocalFont(fontName);
+  if (await byteData == null) {
+    byteData = _httpFetchFont(fontName, fontUrl);
+  }
+  fontLoader.addFont(byteData);
+  await fontLoader.load();
+  print('loaded $fontName');
```
Comment (jacob314) [inline google_fonts/lib/src/google_fonts_base.dart]: remove print before landing?

File: google_fonts/lib/src/google_fonts_base.dart [hunk 6]
```diff
@@ -0,0 +1,190 @@
...
+
+String _fontWeightString(TextStyle textStyle) {
+  if (textStyle.fontWeight == null || textStyle.fontWeight == FontWeight.w400) {
+    return '';
+  }
+  return textStyle.fontWeight.toString().replaceAll('FontWeight.w', '');
+}
+
+String _fontStyleString(FontStyle fontstyle) {
+  if (fontstyle == FontStyle.italic) {
```
Comment (jacob314) [inline google_fonts/lib/src/google_fonts_base.dart]: nit:
```dart
return fontsTyle == FontStyle.italic ? 'italic' : '';
```
is probably cleaner.

File: google_fonts/lib/src/google_fonts_base.dart [hunk 7]
```diff
@@ -0,0 +1,190 @@
...
+  }
+  int score = (a.fontWeight - b.fontWeight).abs();
+  if (a.fontStyle != b.fontStyle) {
+    score += 2;
+  }
+  return score;
+}
+
+const _regular = 'regular';
+const _italic = 'italic';
```
Comment (jacob314) [inline google_fonts/lib/src/google_fonts_base.dart]: use or remove this constants.

File: google_fonts/lib/src/google_fonts_base.dart [hunk 8]
```diff
@@ -0,0 +1,190 @@
...
+  // https://github.com/flutter/flutter/issues/44460
+  PaintingBinding.instance.handleSystemMessage({'type': 'fontsChange'});
+}
+
+// Returns [GoogleFontsFamily] from `variants` that most closely matches
+// [style] according to the computeMatch scoring function.
+//
+// This logic is taken from the following section of the minikin library, which
+// is ultimately how flutter handles matching fonts.
+// * https://github.com/flutter/engine/blob/master/third_party/txt/src/minikin/FontFamily.cpp#L149
```
Comment (jacob314) [inline google_fonts/lib/src/google_fonts_base.dart]: nit: remove the `* `

File: google_fonts/lib/src/README.md (hunk 2 above)
Comment (clocksmith) [reply google_fonts/lib/src/README.md]: Just running the dart script (step 2). That will create the generated output file

File: google_fonts/lib/src/google_fonts_base.dart (hunk 7 above)
Comment (clocksmith) [reply google_fonts/lib/src/google_fonts_base.dart]: They are being used in GoogleFontsFamily.fromString

File: google_fonts/lib/src/google_fonts_base.dart [hunk 9]
```diff
@@ -0,0 +1,188 @@
...
+
+String _fontWeightString(TextStyle textStyle) {
+  if (textStyle.fontWeight == null || textStyle.fontWeight == FontWeight.w400) {
+    return '';
+  }
+  return textStyle.fontWeight.toString().replaceAll('FontWeight.w', '');
+}
+
+String _fontStyleString(FontStyle fontStyle) {
+  return fontStyle == FontStyle.italic ? 'italic' : '';
```
Comment (jacob314) [inline google_fonts/lib/src/google_fonts_base.dart]: use the _italic constant here

**Pull request merged**